### DISCUSSION TITLE
Probe Assisted Tramming.  How's that work?

### DISCUSSION THREAD

**Discussion by grilparto** [Jyers/Marlin#1083] [General Discussion]
Was trying out one of the new commits that added Probe Assisted Tramming.  On the surface, I guess it's working.  Goes to each screw corner (or as close as possible) and probes each corner.  Then you can manually probe each corner and have it relay how many mm high or low it is and adjust/re-probe.  

Curious how this is working though.  Is this like doing a G35 command?  Or something different?  

When it gives you a result that shows x-mm high or low, what is that in relation to?  The height it probed for the center?  I thought G35 produced results for each of the other 3 corners in relation to the bottom left.  If it's the former, is it best to do the probe assisted tramming after building a mesh with ABL or UBL?  Before or after setting the z-offset?  Just curious if I'm using it properly.

**Comment by smooreace** [comment]
Interesting... Ill be keeping an eye on this!

Z offset is a concrete number that will never change unless you change
either your probe mounting or nozzle. So it doesn't matter when its set.
Mine is defined in the firmware with my other probe offset figures. One
less thing to have to worry about!

**Comment by Reggi3** [comment]
I've used it and it was relatively straightforward.  I'm going by my small sample set of using it one time with my e3v2 already setup with a decent 5x5 ABL mesh and no adhesion issues, along with Z offset being well adjusted.  I found that it was really easy to get decent numbers, I think I got down to around 0.005mm for all 4 points, which also translated well when I did a 5x5 mesh afterwards, probably the best mesh I've had and the deviation between high and low points is very much reduced, I hadn't expected that bonus outcome from it.

As I understand things, it creates a simple 4 point mesh of the 4 bed adjustment points, I had a quick look at the code, it seems to average the corner heights and do everything from that as a reference point, like G35 but I don't think it's got a preferred corner.  I'm hoping @Jyers will clarify this for us...

Personally, the way I thought it should be used is before you've set anything else up, this is supposed to replace manual levelling, so on a stock setup machine with JyersUI self compiled firmware installed and that would be before Z offset and any Manual/ABL/UBL meshes have been done.  I think probe assisted levelling is the feature that everyone expects their BLTouch to do, even if you do have to adjust the knobs yourself.

However, as you can see from my experience, if you have already setup your machine, the routine doesn't seem to care and the outcome is better than I've managed un-assisted.  I think no matter when you do the probe assisted levelling, the key is to do a mesh once you've done the probe assist routine and you will always be in the best place to start a print.

I also envisage it being used if you're a beginner/tinkerer/modder, any time you teardown the Y carriage or adjust eccentric nuts, you're going to have to re-tram the bed, anytime you add new springs, lock nuts, silicone inserts or adjustment knobs, again, you're going to be re-tramming.  Probe assisted levelling will streamline the process of getting up and running.  

Jyers goes paperless :-D

**Comment by Jyers** [comment]
@Reggi3 yup, that's about it! The code is my own, no gcode involved. It probes as close as it can to all the bed screws and takes the average of those values. It then uses that average as your reference point when probing the points manually. This should result in the minimal change in overall bed height and the least adjustment needed in order to level the bed.

**Reply by Reggi3** [reply]
It works incredibly well imho  I've wanted a feature like this for a while and I wasn't disappointed, Thank You!

I would 100% recommend that if you can compile your own firmware, you should definitely compile the latest jyersUI code and give this feature a go and let us know how you get on with it.  Once this hits the releases, I think it's going to be a very, very popular feature.

**Reply by GordinhoPl** [reply]
Good to know. I saw this feature when I was building my own firmware, but finally I decided to install prebuild firmware and forgot about that. 
So this feature is disabled in precombiled firmware?

**Reply by Reggi3** [reply]
It's not disabled, just that when the 1.3.4 firmwares were compiled, that feature didn't exist at that point in time afaik :-)

**Reply by GordinhoPl** [reply]
Hmm... I use 1.3.4 precompiled release E3V2-BLTouch-5x5 and printer does't recognize command G35. But I saw this option to enable when trying to build my own compilation, so this is why I think this isn't enabled in precompiled releases

**Reply by Reggi3** [reply]
As previously stated by Jyers, it's not Gcode, so it's not G35, it's Jyers own code, which did not exist when the 1.3.4 release was released.

There isn't anything you actually need to do in the configs to make this feature work, just compile from the latest source code, upload it and
1. The feature will be present in the levelling menu. 
2. Select the 'Probe' tickbox, the printer will home. 
3. You will be presented with a choice of corners, select a corner. 
4. Wait for it to probe, keep an eye on the LCD for feedback for how high/low that corner is.
5. Turn the appropriate adjustment knob, then probe again, repeat until you're happy with the numbers.  
6. Once you've done the first corner, repeat steps 3-5 for the remaining corners.
7. Now do your favourite mesh levelling routine (Manual, ABL or UBL) and check the numbers/mesh.
8. That's it, you should be smiling at this point, enjoy your printer :-)

**Reply by GordinhoPl** [reply]
So i don't need to change anything in config, even I don't need to uncomment ASSISTED_TRAMMING in Configuration_adv.h, as shown below?
![2021-06-29_22h47_26](https://user-images.githubusercontent.com/85198501/123869745-afbdab80-d931-11eb-9c5f-f8afb6895b73.png)

**Reply by Reggi3** [reply]
Last time for the people at the back ;-) There is nothing that you need to change or enable in the configuration.h/configuration_adv.h files, all you need to do is compile the latest version of the source code and the feature will be available for you.  No Gcode was harmed whilst answering this question.

**Reply by GordinhoPl** [reply]
Ok, thanks than and sry for a lot of questions ;)

**Reply by darco415** [reply]
thats some great advice @reggi3.....Im confused w/these steps? What numbers am I shooting for to make me happy? I see the numbers the LCD is showing, but dont know which direction to go. Am I just trying to make each corner the same number? Wait thats the answer right?

4.Wait for it to probe, keep an eye on the LCD for feedback for how high/low that corner is.
5.Turn the appropriate adjustment knob, then probe again, repeat until you're happy with the numbers.

**Reply by Reggi3** [reply]
Well, imho, you're attempting to level the 4 points to each other (and also the X gantry as that is the axis the probe is attached to), so the closer you can get to zero, the better.  In my case I stopped chasing it at 0.005mm, I think that's approaching the limit of the probe but on at least one other knob I got 0.003mm, the numbers are probably excessive but easy to achieve!  

Imho, I think that everyone has a slightly different printer and slightly different tolerance levels for achieving a good first layer, 0.100mm or below is a reasonable target for me using the paper method, I'm guaranteed to get a first layer down without issues.  

If you look at the difference in values I was happy with (0.1mm) using the paper method vs the values I got from the new feature (0.005mm) you can see I'm getting 20x the precision, so the numbers that make you happy will come from your own experience of how well you know your machine and the numbers you get from probing a mesh.

My usual routine is to use octoprint bed visualiser plugin to recall the last mesh, after that I will heat the bed and then do a 5x5 ABL mesh, once that has finished, I will compare the numbers from the new mesh to the old one.  I expect the numbers to be fairly close, if they're not it's an indication that I might have a mechanical problem with the Y carriage (rollers, nuts, something needs adjusting/replacing).  If the numbers are close (within my own arbitrary tolerance range) then it's time to print.

As for which direction, when it probes it will put a message on the LCD with the value and say whether it's high or low, if it says 0.075 High, you should adjust the knob so that it moves lower (turn anti-clockwise).  The great thing about getting feedback on the LCD is that you can turn the knob, re-probe the point and immediately see if you turned the knob in the correct direction.

Another tip for anyone using this feature (or levelling in general) is a 5.14 degree turn of the knob in either direction will change the bed height by 0.01mm, as the adjustment bolt gives us 0.70mm per revolution, 360 degrees/70 gives us a good idea of how far to turn the knob, so if you want to hit single micron adjustments, you need to turn the knob between 0.5 and 1 degree.

@GordinhoPl No problem, it's always good to ask questions, I was just running out of different ways to answer.

**Comment by GordinhoPl** [comment]
Hi. 
About 2 weeks ago I compiled my firmware from source code in "Releases" section and like @Reggi3  told me I didn't need to change anything and by default in Prepare->Manual Leveling menu I had option "Use probe". 
Today I downloaded and tried to build firmware again from fresh source code, and option "Use probe" disappeared from Manual Leveling menu. 
So, now this tool isn't enabled by default and I need to enable this option in Configuration_adv.h file?
Thanks for help :)

**Reply by Cal-LF** [reply]
@GordinhoPl I think it is not working in the latest release. I believe @Jyers is aware of this.

**Reply by darco415** [reply]
all seems good after latest update

**Reply by GordinhoPl** [reply]
So if I enabled G35 command option in Configuration_adv.h this feature will work now? Am I right?

**Reply by Flexigem** [reply]
Does one compile the source from the releases section or the files on the git itself? I think latter is up to date and carries the changes, doesn’t it?
What’s the difference between the jyers and E3V2 configuration files?

**Reply by GordinhoPl** [reply]
I tried both from releases section and git itself. In both there is no option "Use probe"

**Reply by Jyers** [reply]
If you download the latest source code from GitHub and compile that the use probe option should be there, no need to enable anything

**Reply by GordinhoPl** [reply]
@Jyers thanks for you reply, it is strange, because yesterday i tried to do this and I don't have this option. Will try again today. So my last question, the source code from GitHub branch is the same source code then in releases section? Or which one is the most actual?

**Reply by Jyers** [reply]
Downloading the code from the code tab will give you the most recent source which using the zip file provided with the releases will give you the source code at the time of that release.
